### PR TITLE
Cloudfront Invalidator

### PR DESCRIPTION
**Pull request by Bento007**
Invalidate the cloudfront distribution when changes are detected in the S3 Bucket hosting the sight. This makes the website changes immediately visible.

The name of the new lambda is `browser-frontend-$DEPLOYMENT_STAGE-update-site`

### CODE REVIEW

File: browser/frontend/lambdas/Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+SHELL=/bin/bash -o pipefail
+tests:=$(wildcard tests/test*.py)
+
+ifndef APP_NAME
+$(error Please set APP_NAME parameter using `make target APP_NAME=foo)
```
Comment (calvinnhieu) [inline browser/frontend/lambdas/Makefile:5]: missing `

File: browser/frontend/lambdas/chalice/browser-frontend/Makefile [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+endif
+export EXPORT_ENV_VARS_TO_LAMBDA=APP_NAME DEPLOYMENT_STAGE
+
+.PHONY: clean
+clean:
+	rm -rf dist .chalice/deployments chalicelib .chalice/config.json .chalice/policy-*
+
+.PHONY: destroy
+destroy:
+	chalice delete --stage $(DEPLOYMENT_STAGE
```
Comment (calvinnhieu) [inline browser/frontend/lambdas/chalice/browser-frontend/Makefile:17]: missing )

File: browser/frontend/lambdas/chalice/browser-frontend/Makefile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+.PHONY: package
+package: build-chalice-config
+	mkdir -p chalicelib/code
+	cp -R ../../code chalicelib
+
+.PHONY: deploy
+deploy: package
+	chalice deploy --stage $(DEPLOYMENT_STAGE) --api-gateway-stage $(DEPLOYMENT_STAGE)
```
Comment (calvinnhieu) [inline browser/frontend/lambdas/chalice/browser-frontend/Makefile:26]: Is `--api-gateway-stage` still required?

File: browser/frontend/lambdas/code/cloudfront.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+cloudfront_client = boto3.client("cloudfront")
+
+
+def get_cloudfront_distribution(bucket_name: str) -> List[str]:
+    cf_distributions = cloudfront_client.list_distributions()["DistributionList"]["Items"]
+    distributions = []
+    for d in cf_distributions:
+        for o in d["Origins"]["Items"]:
+            if bucket_name in o["DomainName"]:
```
Comment (calvinnhieu) [inline browser/frontend/lambdas/code/cloudfront.py:14]: Can we avoid single letter variables for clarity?

File: tests/unit/browser/frontend/lambda/code/test_cloudfront.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            }
+        }
+        with self.subTest("Single distribution"):
+            self.assertListEqual(get_cloudfront_distribution("test_bucket-one-match"), ["id1"])
+        with self.subTest("Multiple distributions"):
+            self.assertListEqual(get_cloudfront_distribution("test_bucket"), ["id1", "id2"])
+        with self.subTest("No Match"):
+            self.assertListEqual(get_cloudfront_distribution("no-match"), [])
+
+    @mock.patch("browser.frontend.lambdas.code.cloudfront.cloudfront_client" ".create_invalidation")
```
Comment (calvinnhieu) [inline tests/unit/browser/frontend/lambda/code/test_cloudfront.py:24]: Does this work?

File: browser/frontend/lambdas/code/cloudfront.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+                distributions.append(d["Id"])
+    return distributions
+
+
+def invalidate_distributions(distributions: List[str]) -> List[dict]:
+    responses = []
+    for d in distributions:
+        timestamp = datetime.utcnow().replace(second=0, microsecond=0).strftime("%Y%m%d%H%M%S")
+        # Limit the number of invalidation request to 1 per minute per distribution.
```
Comment (calvinnhieu) [inline browser/frontend/lambdas/code/cloudfront.py:23]: Could you explain how this is enforced?

File: tests/unit/browser/frontend/lambda/code/test_cloudfront.py (hunk 5 above)
Comment (Bento007) [reply tests/unit/browser/frontend/lambda/code/test_cloudfront.py:24]: it does work, but that was "black" messing up my formatting. I fixed it.

File: browser/frontend/lambdas/code/cloudfront.py (hunk 6 above)
Comment (Bento007) [reply browser/frontend/lambdas/code/cloudfront.py:23]: Adding to function documentation

File: browser/frontend/lambdas/chalice/browser-frontend/Makefile (hunk 3 above)
Comment (Bento007) [reply browser/frontend/lambdas/chalice/browser-frontend/Makefile]: Nope, I'll remove

**Comment by Bento007** [comment]
> I think logically it makes more sense for the lambdas dir to sit at browser/lambdas. Thoughts?

I'm fine with this. Since this code only affects the frontend I tried to keep all of the relevant code group together.  My next plan was to move the contents of frontend related to the website to a folder call website. so `frontend` would contain `lambdas` and `website`

> Is the lambdas/chalice dir necessary?
I did this incase we add more lambda functions in the future. They can be placed here and managed in an organized way.

File: browser/lambdas/chalice/cloudfront-invalidator/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
+# Cloudfront-Invalidator
+
+A chalice application used to invalidate the cache of a cloudfront distribution for
+an s3 bucket hosting a website. When a change is detected in the s3 bucket, the
+AWS Cloudfront distribution for that bucket is invalidated. This allowschanges
```
Comment (calvinnhieu) [inline browser/lambdas/chalice/cloudfront-invalidator/README.md:5]: space between allows changes

**Pull request merged**